### PR TITLE
Using index mappers instead of own created storage for states of filtering

### PR DESCRIPTION
**Pull request by wszymanski**
### Context
PR https://github.com/handsontable/handsontable/pull/7276 introduced a new kind of Index Map. I'm using it here for purpose of storing and managing states of filtering for particular columns. The `Filters` plugin has been adjusted to use the only Index Map (all unnecessary parts of code has been removed). What's more:

- I removed `clearConditions` method of `ConditionCollection` element. There was a function that cleared conditions at the specified column index but without clearing stack order (stored in separate array). Since the order of columns is stored inside the Index Mapper there is no way to threaten it separately. Thus, some parts of the code have been changed. Hooks `beforeClear` and `afterClear` have been removed while `afterRemove` hook has been introduced.
- I changed `isMatchInConditions` method of `ConditionCollection` element. It will need an argument. Function without an argument hasn't been used internally.
- I introduced an extra method - named `clearValue` - for newly created Index Map within https://github.com/handsontable/handsontable/pull/7276.
- The internal "hiddenRowsCache" map was removed as it turned out that is not necessary at all. The UI components' hidden state can be calculated on the fly while showing the Dropdown
menu.
 - The Filters UI components previously used the `stateSaver` mixing which causes the components state misalignment after column reordering. The mixin is removed and the state management from now on is implemented directly into the UI Base component.

### How has this been tested?

Manual tests for alter actions.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):
1. #6830
2. #6397

### Checklist:

- [x] My code follows the code style of this project,
- [ ] My change requires a change to the documentation.

### CODE REVIEW

File: src/translations/maps/linkedPhysicalIndexToValueMap.js [hunk 1]
```diff
@@ -75,6 +76,22 @@ class LinkedPhysicalIndexToValueMap extends IndexMap {
     super.clear();
   }
 
+  /**
+   * Remove value for particular index.
+   *
+   * @param {number} physicalIndex Physical index.
+   */
+  clearValue(physicalIndex) {
```
Comment (budnix) [inline src/translations/maps/linkedPhysicalIndexToValueMap.js:84]: I haven't found any unit test for that method. Could I ask you to cover that?

**Review by budnix — CHANGES_REQUESTED** [review]
You changed fewer lines of code than I originally assumed 👍  Good Job! I left some observations and comments in the source code.

File: src/plugins/filters/filters.js [hunk 2]
```diff
@@ -194,12 +197,13 @@ class Filters extends BasePlugin {
       })));
     }
     if (!this.conditionCollection) {
-      this.conditionCollection = new ConditionCollection();
+      this.conditionCollection = new ConditionCollection(this.sortingStates);
```
[the remark this answers is not in the document]
Comment (wszymanski) [reply src/plugins/filters/filters.js:200]: That's the right direction, but there is a problem. Created index mappers within the mentioned PR are also registered. I'm afraid, that there should be no registration for one use case.

https://github.com/handsontable/handsontable/blob/46c62ae1d0593fe2b4f49f5f3bcf6dbf1830cb9b/src/plugins/filters/conditionUpdateObserver.js#L139

This temporary collection isn't registered to the IndexMapper. WDYT?

File: src/translations/maps/linkedPhysicalIndexToValueMap.js (hunk 1 above)
Comment (wszymanski) [reply src/translations/maps/linkedPhysicalIndexToValueMap.js:84]: The `setValueAtIndex` already removes a value from the mentioned array.

https://github.com/handsontable/handsontable/blob/8ee91484fe1efa000c577faadbcdc11fac0d8f87/src/translations/maps/linkedPhysicalIndexToValueMap.js#L54-L60

However, the method does extra execution of `initValueOrFn` function or pass initial value to the `setValueAtIndex`.

I've used `clearValue` name as it works very similar to the current `clear` method.

File: src/plugins/filters/filters.js [hunk 3]
```diff
@@ -139,6 +139,9 @@ class Filters extends BasePlugin {
 
     this.filtersRowsMap = this.hot.rowIndexMapper.registerMap(this.pluginName, new TrimmingMap());
     this.dropdownMenuPlugin = this.hot.getPlugin('dropdownMenu');
+    this.sortingStates = this.hot.columnIndexMapper.registerMap(
```
[the remark this answers is not in the document]
Comment (wszymanski) [reply src/plugins/filters/filters.js]: Changed within https://github.com/handsontable/handsontable/pull/7319/commits/c7eb3b9c7e3acfb339fcbdf836c998bf80a4e108.

File: src/plugins/filters/conditionCollection.js [hunk 4]
```diff
@@ -166,20 +153,13 @@ class ConditionCollection {
...
    */
   exportAllConditions() {
-    const result = [];
-
-    arrayEach(this.orderStack, (column) => {
-      const conditions = arrayMap(this.getConditions(column), ({ name, args }) => ({ name, args }));
-      const operation = this.columnTypes[column];
-
-      result.push({
+    return arrayReduce(this.filteringStates.getEntries(), (allConditions, [column, { operation, conditions }]) => {
```
[the remark this answers is not in the document]
Comment (wszymanski) [reply src/plugins/filters/conditionCollection.js:157]: Changed within https://github.com/handsontable/handsontable/pull/7319/commits/c7eb3b9c7e3acfb339fcbdf836c998bf80a4e108.

**Comment by wszymanski** [comment]
> You changed fewer lines of code than I originally assumed 👍 Good Job! I left some observations and comments in the source code.
> 
> Additionally, I've found that the TS definition wasn't updated yet. Mostly about `ConditionCollection` class.

Thanks. Documentation should be also updated.

File: src/translations/maps/linkedPhysicalIndexToValueMap.js (hunk 1 above)
Comment (wszymanski) [reply src/translations/maps/linkedPhysicalIndexToValueMap.js:75]: Added test within https://github.com/handsontable/handsontable/pull/7319/commits/c4091ccc7d255d07298631285aa5b3c8aa563b7d.

**Comment by budnix** [comment]
I think that I've found a bug. When you filter columns using the "filter_by_value" component the data within that component don't react after changing the conditions from the previous columns.

Before the PR:
beforePR

Within the PR
afterPR

File: src/plugins/filters/filters.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (budnix) [reply src/plugins/filters/filters.js]: Maybe we could add an option to the `ConditionCollection` class wherein the constructor the map registration will be differentiated according to the existence of the option.

So the constructor signature could look something like:
```
new ConditionCollection(hot, { trackIndexes: true });
```
PS: I believe that you would find a better name for the `trackIndexes` option.

**Comment by wszymanski** [comment]
Within this PR I removed `clearConditions` method of ConditionCollection element. There was a function that cleared conditions at the specified column index but **without clearing stack order** (stored in a separate array). Since the order of columns is stored inside the Index Mapper there is no way to threaten it separately in a clean way.

Bug found by @budnix was related to the fact that after refactor, the drop-down menu removed (not only "temporary cleared" value) all conditions for a particular column (they may be overwritten). Thus, the filtered column was completely removed from the queue of filtering.

https://github.com/handsontable/handsontable/blob/9375365c5f566c6e829ce252069de68661285768/src/plugins/filters/filters.js#L613-L621

In the next step, it could be added to the queue but **not in the previous position** but at the last position.

https://github.com/handsontable/handsontable/blob/9375365c5f566c6e829ce252069de68661285768/src/plugins/filters/filters.js#L623-L633

I've added an extra argument to the `setValueAtIndex` of the newly created index map.

https://github.com/handsontable/handsontable/blob/63c22ace8677646d7c74b6cc3f96c6805ed9412e/src/translations/maps/linkedPhysicalIndexToValueMap.js#L55

It should fix the problem. I've prepared also changes similar to this suggested by @budnix within [a comment](https://github.com/handsontable/handsontable/pull/7319#discussion_r508262099). A way how we create states manager for the `ColumnSorting` and `Filters` plugins is very similar from now.

PS. I've tried to add one more extra argument to the `setValueAtIndex` with quite unnatural default. I was looking for a way how to move part of `clearValue` method to the `Index Map`. However, it looked quite quirky and I've decided to remove it.

File: src/translations/maps/linkedPhysicalIndexToValueMap.js [hunk 5]
```diff
@@ -47,15 +48,16 @@ class LinkedPhysicalIndexToValueMap extends IndexMap {
    *
    * @param {number} index The index.
    * @param {*} value The value to save.
+   * @param {number} position Position to which entry will be added.
    *
    * @returns {boolean}
    */
-  setValueAtIndex(index, value) {
+  setValueAtIndex(index, value, position = this.orderOfIndexes.length) {
```
Comment (budnix) [inline src/translations/maps/linkedPhysicalIndexToValueMap.js:55]: I haven't seen unit tests for the new argument `position`. Can I ask you to add it?

**Review by budnix — CHANGES_REQUESTED** [review]
~🤔  I've tried the use case https://github.com/handsontable/handsontable/pull/7319#issuecomment-712164184 using the newest changes but It seems that there is still a bug. Could you check that?~

I've checked again and I haven't found any issues.

File: src/plugins/filters/conditionUpdateObserver.js [hunk 6]
```diff
@@ -16,7 +16,7 @@ import { createArrayAssertion } from './utils';
  * @plugin Filters
  */
 class ConditionUpdateObserver {
-  constructor(conditionCollection, columnDataFactory = () => []) {
+  constructor(conditionCollection, columnDataFactory = () => [], hot) {
```
[the remark this answers is not in the document]
Comment (budnix) [reply src/plugins/filters/conditionUpdateObserver.js]: Fixed within https://github.com/handsontable/handsontable/pull/7319/commits/3356ad8640444fc0e564dda152d58350d6b1d92c

File: src/translations/maps/linkedPhysicalIndexToValueMap.js (hunk 5 above)
Comment (budnix) [reply src/translations/maps/linkedPhysicalIndexToValueMap.js:55]: Fixed within https://github.com/handsontable/handsontable/pull/7319/commits/9e403c163b67182122c4708f59abfaf3dd1efd11

**Comment by budnix** [comment]
While testing the "by_value" component I've found another bug where the content of the component is not refreshed after columns reorder.
Kapture 2020-11-10 at 09 51 01

**Review by swistach — COMMENTED** [review]
Shouldn't we restore possible values after remove column with active `filter_by_value`?
altering_columns_filters

**Comment by budnix** [comment]
> Shouldn't we restore possible values after remove column with active `filter_by_value`?
> altering_columns_filters

Filters are passive, so I'd say no. You can call the `hot.filter()` after column removal to achieve active-like behavior.

File: src/plugins/filters/conditionCollection.js [hunk 7]
```diff
@@ -1,37 +1,48 @@
...
     /**
-     * Conditions collection grouped by operation type and then column index.
+     * Handsontable instance.
      *
-     * @type {object}
+     * @type {Core}
      */
-    this.conditions = this.initConditionsCollection();
-
+    this.hot = hot;
```
[the remark this answers is not in the document]
Comment (budnix) [reply src/plugins/filters/conditionCollection.js:24]: Fixed within https://github.com/handsontable/handsontable/pull/7319/commits/f4cc1522265b42a7540858ae814ef51e35be358f

**Review by swistach — APPROVED** [review]

**Review by wojciechczerniak — APPROVED** [review]
🥇

**Pull request merged**